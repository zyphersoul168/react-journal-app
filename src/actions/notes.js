import Swal from "sweetalert2";

import { addDoc, deleteDoc, setDoc, collection, doc } from "firebase/firestore";
import { db } from "../firebase/config";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    const userRef = collection(db, `${uid}/journal/notes`);

    try {
      // add note to database
      const noteRef = await addDoc(userRef, newNote);

      dispatch(activeNote(noteRef.id, newNote));
      dispatch(addNewNote(noteRef.id, newNote));
    } catch (error) {
      console.log(error);
    }
  };
};

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id,
    ...note,
  },
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }

    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    await setDoc(doc(db, `${uid}/journal/notes`, note.id), noteToFirestore);

    dispatch(refreshNote(note.id, noteToFirestore));
    Swal.fire("Saved!", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdate,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUploadingImage = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;

    Swal.fire({
      title: "Uploading...",
      text: "Wait :)",
      allowOutsideClick: false,
      didOpen: () => {
        Swal.showLoading();
      },
    });

    const fileUrl = await fileUpload(file);
    activeNote.url = fileUrl;

    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const startDeletingNote = (noteId) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    await deleteDoc(doc(db, `${uid}/journal/notes`, noteId));

    dispatch(deleteNote(noteId));

    Swal.fire("Deleted!", "Note deleted", "success");
  };
};

export const deleteNote = (noteId) => ({
  type: types.notesDelete,
  payload: noteId,
});

export const cleanNotes = () => ({
  type: types.notesLogoutClean,
});
