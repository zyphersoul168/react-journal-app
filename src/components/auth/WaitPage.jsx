import React from 'react';

export const WaitPage = () => {
  return (
    <div className="wait__container">
      <h1>{'Wait... :)'}</h1>
    </div>
  );
};
