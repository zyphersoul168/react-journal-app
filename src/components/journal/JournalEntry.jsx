import React from 'react';

export const JournalEntry = () => {
  return (
    <div className="journal__entry pointer">
      <div
        className="journal__entry-picture"
        style={{
          backgroundSize: 'cover',
          backgroundImage:
            'url(https://pm1.narvii.com/7673/bc18bebccadba969afa1b0778c53bbdfa93bda26r1-540-553v2_hq.jpg)',
        }}></div>
      <div className="journal__entry-body">
        <p className="journal__entry-title">Un nuevo dÃ­a</p>
        <p className="journal__entry-content">
          Cillum tempor est esse incididunt ex velit est quis pariatur
          consectetur.
        </p>
      </div>
      <div className="journal__entry-date-box">
        <span>Monday</span>
        <h4>5</h4>
      </div>
    </div>
  );
};
