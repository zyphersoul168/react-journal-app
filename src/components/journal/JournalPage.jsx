import React from 'react'

export const JournalPage = () => {
  return (
    <div>
      <h1>JournalPage</h1>
    </div>
  )
}
