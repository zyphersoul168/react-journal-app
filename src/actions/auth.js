import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut,
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/config';
import { types } from '../types/types';
import { finishLoading, startLoading } from './ui';
import Swal from 'sweetalert2';
import { errors } from '../types/errors';
import { cleanNotes } from './notes';

const auth = getAuth();

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    dispatch(startLoading());
    try {
      const { user } = await signInWithEmailAndPassword(auth, email, password);
      dispatch(login(user.uid, user.displayName));
      dispatch(finishLoading());
    } catch (error) {
      dispatch(finishLoading());
      error.message === errors.userNotFound &&
        Swal.fire('Error', 'User not found', 'error');
      error.message === errors.wrongPassword &&
        Swal.fire('Error', 'Wrong Password', 'error');
    }
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((error) => {
        error.message === errors.existingAccount &&
          Swal.fire('Error', 'There is already an account with this email', 'error');
        console.log(error.message);
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    await signOut(auth);
    dispatch(logout());
    dispatch(cleanNotes());
  };
};

export const logout = () => ({
  type: types.logout,
});
