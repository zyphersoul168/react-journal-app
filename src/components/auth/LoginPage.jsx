import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginPage = () => {
  const dispatch = useDispatch();
  const { loading } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: 'ortiz@mail.com',
    password: 'aaroneduardo1',
  });

  const { email, password } = formValues;

  const handleLogin = (event) => {
    event.preventDefault();
    dispatch(startLoginEmailPassword(email, password));
  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <div>
      <h1>Log in</h1>
      <form onSubmit={handleLogin}>
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          value={password}
          onChange={handleInputChange}
        />
        <button disabled={ loading } className="auth__button">
          <span className="auth__login">Log in</span>
        </button>
        <hr />
        <div>
          <span>Or you can log in with:</span>
          <div className="google-btn" onClick={handleGoogleLogin}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Continue with google</b>
            </p>
          </div>
        </div>
        <span>
          Don't have an account?{' '}
          <Link className="link" to="/auth/register">
            Sign up
          </Link>
        </span>
      </form>
    </div>
  );
};
