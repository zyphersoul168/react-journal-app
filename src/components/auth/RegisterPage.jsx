import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import { removeError, setError } from '../../actions/ui';
import { useForm } from '../../hooks/useForm';

export const RegisterPage = () => {
  const dispatch = useDispatch();
  const { msgError } = useSelector((state) => state.ui);

  const [formValues, handleInputChange] = useForm({
    name: 'aaron',
    email: 'ortiz@mail.com',
    password: '1234',
    password2: '1234',
  });

  const { name, email, password, password2 } = formValues;

  const handleRegister = (event) => {
    event.preventDefault();
    isFormValid() && dispatch(startRegisterWithEmailPasswordName(email, password, name))
  };

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required'));
      return false;
    } else if (!validator.isEmail(email)) {
      dispatch(setError('Not valid email'));
      return false;
    } else if (password !== password2 || password.length < 5) {
      dispatch(setError('Password too short'));
      return false;
    }

    dispatch(removeError());
    return true;
  };

  return (
    <div>
      <h1>Sign up</h1>
      <form onSubmit={handleRegister}>
        {msgError && <div className="auth__alert">{msgError}</div>}
        <input
          className="auth__input"
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="on"
          onChange={handleInputChange}
          value={name}
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          onChange={handleInputChange}
          value={email}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Password"
          name="password"
          onChange={handleInputChange}
          value={password}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Confirm password"
          name="password2"
          autoComplete="off"
          onChange={handleInputChange}
          value={password2}
        />
        <button className="auth__button">
          <span className="auth__login">Sign up</span>
        </button>
        <hr />
        <div>
          <span>Or you can sign up with:</span>
          <div className="google-btn">
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Continue with google</b>
            </p>
          </div>
        </div>
        <span>
          Already have an account?{' '}
          <Link className="link" to="/auth/login">
            Log in
          </Link>
        </span>
      </form>
    </div>
  );
};
