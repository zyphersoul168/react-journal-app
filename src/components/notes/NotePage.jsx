import React, { useEffect, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import { NotesAppBar } from "./NotesAppBar";
import { useForm } from "../../hooks/useForm";
import { activeNote } from "../../actions/notes";

export const NotePage = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector((state) => state.notes);
  const [formValues, handleInputChange, reset] = useForm(note);
  const { body, title } = formValues;
  const activeId = useRef(note.id);

  useEffect(() => {
    if(note.id !== activeId.current) {
      reset(note);
      activeId.current = note.id
    }
  }, [note, reset]);

  useEffect(() => {
    dispatch(activeNote(formValues.id, {...formValues}))
  }, [formValues, dispatch]);

  return (
    <div className="notes__main-content">
      <NotesAppBar />
      <div className="notes__content">
        <input
          type="text"
          className="notes__title-input"
          placeholder="An awesome history?"
          autoComplete="off"
          name="title"
          value={title}
          onChange={handleInputChange}
        />
        <textarea
          placeholder="Tell it here!"
          className="notes__textarea"
          name="body"
          value={body}
          onChange={handleInputChange}
        ></textarea>
        {note.url && (
          <div className="notes__img">
            <img
              src={note.url}
              alt="imagen"
            />
          </div>
        )}
      </div>
    </div>
  );
};
