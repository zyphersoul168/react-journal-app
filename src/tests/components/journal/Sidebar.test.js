import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import "@testing-library/jest-dom";

import { Sidebar } from "../../../components/journal/Sidebar";
import { startLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = {
  auth: {
    name: "aaron",
  },
  ui: {
    loading: false,
    msgError: null,
  },
  notes: {
    active: {
      id: "abc",
    },
    notes: [],
  },
};
let store = mockStore(initialState);
store.dispatch = jest.fn();

jest.mock("../../../actions/auth", () => ({
  startLogout: jest.fn(),
}));

jest.mock("../../../actions/notes", () => ({
  startNewNote: jest.fn(),
}));

const wrapper = mount(
  <Provider store={store}>
    <Sidebar />
  </Provider>
);

describe("Tests on Sidebar component", () => {
  test("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call logout action", () => {
    wrapper.find(".logout").simulate("click");

    expect(startLogout).toHaveBeenCalled();
  });

  test("should call startNewNote action", () => {
    wrapper.find(".journal__new-entry").simulate("click");

    expect(startNewNote).toHaveBeenCalled();
  });
});
