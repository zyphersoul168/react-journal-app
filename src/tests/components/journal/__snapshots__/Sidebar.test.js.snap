// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on Sidebar component should match snapshot 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-user-circle"
      />
      <span
        className="username"
      >
         aaron
      </span>
    </h3>
    <button
      className="btn logout"
      onClick={[Function]}
    >
      <span>
        Log out
      </span>
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <span>
      <i
        className="far fa-calendar-plus fa-5x"
      />
    </span>
    <p
      className="mt-5"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
